src/groundcrew/tools.py

```diff
@@ -350,3 +370,298 @@ def __call__(self, user_prompt: str, include_code: bool) -> str:
+
+    def __call__(
+            self,
+            user_prompt: str,
+            filepath_inexact: str = 'none',
+            sort_on: str = 'max'
+        ) -> str:
+        """Answer questions using about the complex parts of a codebase.
+        This method will create a summary of the most complex files to help answer
+        the question.
+
+        Args:
+            user_prompt (str): The user's question.
+            filepath_inexact (str): A filepath explicitly requested by the user.  If the
+              user did not explicitly request a filepath, pass 'none'.
+            sort_on (str): Sort the results by the "average" or "max" complexity of the file.
+        """
+
+        if filepath_inexact != 'none':
+            filepath = fuzzy_match_file_path(self.collection, filepath_inexact, 50)
+
+            if filepath is None:
+                return f'Could not find a source file matching `{filepath_inexact}`'
+
+            all_items = self.collection.get(
+                include=['metadatas'],
+                where={'filepath': filepath}
+            )
+
+            files = {
+                id_: metadata['text']
+                for id_, metadata in zip(all_items['ids'], all_items['metadatas'])
+                if id_[-3:] == '.py'
+            }
+
+        else:
+            files = get_python_files(self.collection)
+
+        sort_on = sort_on.lower()
+        if sort_on not in ['average', 'max']:
+            return 'The sort_on parameter passed to CyclomaticComplexityTool must be "average" or "max".'
+
+        complexity_summary_str = self.complexity_analysis(files, sort_on)
+
+        prompt = (
+            complexity_summary_str +
+            '\n### Task ###\n' + self.base_prompt +
+            '\n### Question ###\n' + user_prompt + '\n'
+        )
+
+        return self.llm(prompt)
+
+    def get_complexity(self, source_code: str) -> dict[str, dict]:
+        """Get the complexity of source code."""
+
+        complexity_dict = cu.cyclomatic_complexity(source_code)
+        if len(complexity_dict) == 0:
+            average_complexity, max_complexity = 0, 0
+        else:
+            average_complexity = sum(complexity_dict[obj]['complexity'] for obj in complexity_dict)/len(complexity_dict)
+            max_complexity = max(complexity_dict[obj]['complexity'] for obj in complexity_dict)
+
+        return {
+            'average': average_complexity,
+            'max': max_complexity,
+            'details': complexity_dict
+        }
+
+    def complexity_analysis(self, files: dict[str, str], sort_on: str) -> tuple[list[str], dict]:
+        """Analyze the complexity of the codebase."""
+
+        file_complexity = {}
+        current_max = 0
+        for file, source_code in files.items():
+            complexity = self.get_complexity(source_code)
+            if complexity['max'] >= self.min_max_complexity or complexity['max'] > current_max:
+                file_complexity[file] = complexity
```

Comment:
Do you want this outside the `if`?

Reply:
Inside the `if` to build up a set of maximally complex objects.  Currently, if an object does not pass the threshold it is not included unless there are no other more complex objects.  I should add to the docstring to clarify this.